File: back/Pregiato.Core/Entities/Message.cs
using System.ComponentModel.DataAnnotations;
using System.Text.Json;

namespace Pregiato.Core.Entities
{
    public enum MessageDirection
    {
        In = 0,
        Out = 1
    }

    public enum MessageType
    {
        Text = 0,
        Image = 1,
        Audio = 2,
        Document = 3,
        Video = 4,
        Voice = 5,       // Nota de voz
        Sticker = 6,     // Figurinha
        Location = 7,    // Localização
        Contact = 8,     // Contato compartilhado
        System = 9       // Mensagem do sistema
    }

    public enum MessageStatus
    {
        Queued = 0,
        Sending = 1,
        Sent = 2,
        Delivered = 3,
        Read = 4,
        Failed = 5
    }

    /// <summary>
    /// Entidade Message Unificada - Suporta todos os tipos de mensagens
    /// Campos obrigatórios mínimos: Id, ConversationId, Direction, Type, CreatedAt, Status
    /// Todos os demais campos são nullable para flexibilidade
    /// </summary>
    public class Message
    {
        // === CAMPOS OBRIGATÓRIOS MÍNIMOS ===
        [Key]
        public Guid Id { get; set; }
        
        [Required]
        public Guid ConversationId { get; set; }  // ChatId do requisito
        
        [Required]
        public MessageDirection Direction { get; set; }
        
        [Required]
        public MessageType Type { get; set; } = MessageType.Text;
        
        [Required]
        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
        
        [Required]
        public MessageStatus Status { get; set; } = MessageStatus.Queued;

        // === IDENTIFICAÇÃO DO REMETENTE ===
        [StringLength(100)]
        public string? SenderId { get; set; }  // ID do usuário/sistema que enviou
        
        [StringLength(50)]
        public string? FromNormalized { get; set; }  // Número normalizado (ex: 5511999999999)
        
        [StringLength(100)]
        public string? FromOriginal { get; set; }   // Número original completo
        
        public bool FromMe { get; set; } = false;   // Enviado por nós ou recebido
        
        public bool IsGroup { get; set; } = false;  // Mensagem de grupo

        // === CONTEÚDO DE TEXTO (NULLABLE) ===
        // Removido StringLength para suportar mensagens longas (Base64 de áudio, etc)
        public string? Text { get; set; }  // Conteúdo da mensagem (texto, legenda)

        // === CAMPOS DE MÍDIA (NULLABLE) ===
        [StringLength(1000)]  // Aumentado para URLs longas
        public string? MediaUrl { get; set; }  // URL da mídia armazenada (S3, drive, etc)
        
        [StringLength(255)]   // Aumentado para nomes longos
        public string? FileName { get; set; }
        
        [StringLength(100)]
        public string? MimeType { get; set; }
        
        public long? Size { get; set; }  // Tamanho do arquivo em bytes
        
        public int? Duration { get; set; }  // Duração em segundos (áudio/vídeo)
        
        [StringLength(1000)]
        public string? Thumbnail { get; set; }  // URL do thumbnail (vídeo/imagem)

        // === CAMPOS DE LOCALIZAÇÃO (NULLABLE) ===
        public decimal? Latitude { get; set; }
        public decimal? Longitude { get; set; }
        [StringLength(500)]
        public string? LocationAddress { get; set; }

        // === CAMPOS DE CONTATO (NULLABLE) ===
        [StringLength(200)]
        public string? ContactName { get; set; }
        [StringLength(50)]
        public string? ContactPhone { get; set; }

        // === METADADOS FLEXÍVEIS ===
        // Removido StringLength para suportar metadados extensos
        public string? Metadata { get; set; }  // JSON para extensões futuras
        
        // Removido StringLength para suportar payload completo original
        public string? PayloadJson { get; set; }  // Payload completo original (backup)

        // === IDENTIFICADORES EXTERNOS ===
        [StringLength(255)]
        public string? ExternalMessageId { get; set; }
        
        [StringLength(255)]
        public string? ClientMessageId { get; set; }
        
        [StringLength(500)]
        public string? WhatsAppMessageId { get; set; }
        
        [StringLength(100)]
        public string? ChatId { get; set; }  // ID do chat/conversa externa

        // === SISTEMA ===
        public Guid? SessionId { get; set; }
        
        [StringLength(1000)]
        public string? InternalNote { get; set; }  // Nota interna do operador
        
        public DateTime? UpdatedAt { get; set; }
        
        // === RELACIONAMENTOS ===
        public virtual Conversation Conversation { get; set; } = null!;
        public virtual ChatSession? Session { get; set; }

        // === PROPRIEDADES COMPUTADAS ===
        public bool HasMedia => !string.IsNullOrEmpty(MediaUrl);
        
        public bool HasLocation => Latitude.HasValue && Longitude.HasValue;
        
        public bool HasContact => !string.IsNullOrEmpty(ContactName);

        public string TypeDescription => Type switch
        {
            MessageType.Text => "Texto",
            MessageType.Image => "Imagem", 
            MessageType.Audio => "Áudio",
            MessageType.Video => "Vídeo",
            MessageType.Voice => "Nota de Voz",
            MessageType.Document => "Documento",
            MessageType.Sticker => "Figurinha",
            MessageType.Location => "Localização",
            MessageType.Contact => "Contato",
            MessageType.System => "Sistema",
            _ => "Desconhecido"
        };

        public string Preview => Type switch
        {
            MessageType.Text => Text?.Substring(0, Math.Min(Text.Length, 100)) ?? "",
            MessageType.Image => "📷 Imagem",
            MessageType.Video => "🎬 Vídeo", 
            MessageType.Audio => "🎵 Áudio",
            MessageType.Voice => "🎤 Nota de Voz",
            MessageType.Document => $"📄 {FileName ?? "Documento"}",
            MessageType.Sticker => "😀 Figurinha",
            MessageType.Location => $"📍 {LocationAddress ?? "Localização"}",
            MessageType.Contact => $"👤 {ContactName ?? "Contato"}",
            MessageType.System => "⚙️ Sistema",
            _ => "❓ Desconhecido"
        };

        // === MÉTODOS UTILITÁRIOS ===
        public void SetMetadata<T>(T data) where T : class
        {
            Metadata = JsonSerializer.Serialize(data);
        }

        public T? GetMetadata<T>() where T : class
        {
            if (string.IsNullOrEmpty(Metadata)) return null;
            try
            {
                return JsonSerializer.Deserialize<T>(Metadata);
            }
            catch
            {
                return null;
            }
        }

        // === COMPATIBILIDADE (DEPRECATED) ===
        [Obsolete("Use Text instead")]
        public string Body 
        { 
            get => Text ?? string.Empty; 
            set => Text = value; 
        }
    }
} 

File: front/src/types/message.ts
/**
 * ============================================================================
 * CONTRATOS UNIFICADOS DE MENSAGENS - ALINHADO COM BACKEND .NET
 * ============================================================================
 * 
 * Esta interface está 100% alinhada com MessageDto do backend (.NET)
 * Suporta todos os tipos de mensagem: texto, mídia, localização, contatos, etc.
 * 
 * IMPORTANTE: Manter sincronização com back/Pregiato.Application/DTOs/MessageDto.cs
 * ============================================================================
 */

// === ENUMS ALINHADOS COM BACKEND ===

export enum MessageDirection {
  In = 0,
  Out = 1
}

export enum MessageType {
  Text = 0,
  Image = 1,
  Audio = 2,
  Document = 3,
  Video = 4,
  Voice = 5,       // Nota de voz
  Sticker = 6,     // Figurinha
  Location = 7,    // Localização
  Contact = 8,     // Contato compartilhado
  System = 9       // Mensagem do sistema
}

export enum MessageStatus {
  Queued = 0,
  Sending = 1,
  Sent = 2,
  Delivered = 3,
  Read = 4,
  Failed = 5
}

// === INTERFACE PRINCIPAL UNIFICADA ===

/**
 * Interface MessageDto unificada - 100% compatível com backend
 * Suporta todos os tipos de mensagem com campos opcionais flexíveis
 */
export interface MessageDto {
  // === CAMPOS OBRIGATÓRIOS ===
  id: string;
  conversationId: string;
  direction: MessageDirection;
  type: MessageType;
  createdAt: string;
  status: MessageStatus;
  fromMe: boolean;
  isGroup: boolean;

  // === IDENTIFICAÇÃO DO REMETENTE ===
  senderId?: string;
  fromNormalized?: string;
  fromOriginal?: string;

  // === CONTEÚDO (NULLABLE) ===
  text?: string;  // Renomeado de Body para Text

  // === MÍDIA (NULLABLE) ===
  mediaUrl?: string;
  fileName?: string;
  mimeType?: string;
  size?: number;
  duration?: number;
  thumbnail?: string;

  // === LOCALIZAÇÃO (NULLABLE) ===
  latitude?: number;
  longitude?: number;
  locationAddress?: string;

  // === CONTATO (NULLABLE) ===
  contactName?: string;
  contactPhone?: string;

  // === METADADOS ===
  metadata?: Record<string, unknown>;  // JSON flexível
  chatId?: string;

  // === IDENTIFICADORES EXTERNOS ===
  externalMessageId?: string;
  clientMessageId?: string;
  whatsAppMessageId?: string;

  // === SISTEMA ===
  sessionId?: string;
  internalNote?: string;
  updatedAt?: string;

  // === CAMPOS COMPUTADOS ===
  hasMedia?: boolean;
  hasLocation?: boolean;
  hasContact?: boolean;
  typeDescription?: string;
  preview?: string;

  // === ANEXOS/ARQUIVOS ===
  attachment?: {
    dataUrl: string;
    mimeType: string;
    fileName?: string;
  } | null;

  // === COMPATIBILIDADE LEGADA (DEPRECATED) ===
  /** @deprecated Use text instead */
  body?: string;
  /** @deprecated Use text instead */
  content?: string;
  /** @deprecated Use createdAt instead */
  ts?: string;
}

// === DTOs ESPECIALIZADOS PARA CRIAÇÃO ===

/**
 * DTO para criação de mensagens de texto
 */
export interface CreateTextMessageDto {
  conversationId: string;
  text: string;
  chatId?: string;
  fromNormalized?: string;
  fromMe?: boolean;
  isGroup?: boolean;
}

/**
 * DTO para criação de mensagens com mídia
 */
export interface CreateMediaMessageDto {
  conversationId: string;
  type: MessageType;
  text?: string;  // Legenda
  mediaUrl: string;
  fileName?: string;
  mimeType?: string;
  size?: number;
  duration?: number;
  thumbnail?: string;
  chatId?: string;
  fromNormalized?: string;
  fromMe?: boolean;
  isGroup?: boolean;
}

/**
 * DTO para criação de mensagens de localização
 */
export interface CreateLocationMessageDto {
  conversationId: string;
  latitude: number;
  longitude: number;
  locationAddress?: string;
  chatId?: string;
  fromNormalized?: string;
  fromMe?: boolean;
  isGroup?: boolean;
}

/**
 * DTO para criação de mensagens de contato
 */
export interface CreateContactMessageDto {
  conversationId: string;
  contactName: string;
  contactPhone: string;
  chatId?: string;
  fromNormalized?: string;
  fromMe?: boolean;
  isGroup?: boolean;
}

/**
 * DTO para atualização de status de mensagem
 */
export interface UpdateMessageDto {
  status?: MessageStatus;
  whatsAppMessageId?: string;
  internalNote?: string;
}

// === DTOs PARA WHATSAPP INCOMING ===

/**
 * DTO para mensagens recebidas do WhatsApp via RabbitMQ
 */
export interface WhatsAppIncomingMessageDto {
  externalMessageId: string;
  from: string;
  fromNormalized: string;
  to: string;
  type: string;
  timestamp: string;
  instanceId: string;
  fromMe: boolean;
  isGroup: boolean;
  body?: string;
  chatId: string;
  attachment?: WhatsAppAttachmentDto;
  location?: WhatsAppLocationDto;
  contact?: WhatsAppContactDto;
  simulated?: boolean;
}

export interface WhatsAppAttachmentDto {
  dataUrl?: string;
  mediaUrl?: string;
  mimeType?: string;
  fileName?: string;
  mediaType?: string;
  fileSize?: number;
  duration?: number;
  width?: number;
  height?: number;
  thumbnail?: string;
}

export interface WhatsAppLocationDto {
  latitude: number;
  longitude: number;
  address?: string;
}

export interface WhatsAppContactDto {
  name: string;
  phone?: string;
}

// === UTILITÁRIOS ===

/**
 * Utilitários para trabalhar com mensagens
 */
export class MessageUtils {
  /**
   * Converte MessageType enum para string legível
   */
  static getTypeDescription(type: MessageType): string {
    switch (type) {
      case MessageType.Text: return 'Texto';
      case MessageType.Image: return 'Imagem';
      case MessageType.Video: return 'Vídeo';
      case MessageType.Audio: return 'Áudio';
      case MessageType.Voice: return 'Nota de Voz';
      case MessageType.Document: return 'Documento';
      case MessageType.Sticker: return 'Figurinha';
      case MessageType.Location: return 'Localização';
      case MessageType.Contact: return 'Contato';
      case MessageType.System: return 'Sistema';
      default: return 'Desconhecido';
    }
  }

  /**
   * Gera preview da mensagem baseado no tipo
   */
  static getPreview(message: MessageDto): string {
    switch (message.type) {
      case MessageType.Text:
        return message.text?.substring(0, 100) || '';
      case MessageType.Image:
        return '📷 Imagem';
      case MessageType.Video:
        return '🎬 Vídeo';
      case MessageType.Audio:
        return '🎵 Áudio';
      case MessageType.Voice:
        return '🎤 Nota de Voz';
      case MessageType.Document:
        return `📄 ${message.fileName || 'Documento'}`;
      case MessageType.Sticker:
        return '😀 Figurinha';
      case MessageType.Location:
        return `📍 ${message.locationAddress || 'Localização'}`;
      case MessageType.Contact:
        return `👤 ${message.contactName || 'Contato'}`;
      case MessageType.System:
        return '⚙️ Sistema';
      default:
        return '❓ Desconhecido';
    }
  }

  /**
   * Verifica se a mensagem tem mídia
   */
  static hasMedia(message: MessageDto): boolean {
    return !!(message.mediaUrl && message.mediaUrl.trim());
  }

  /**
   * Verifica se a mensagem tem localização
   */
  static hasLocation(message: MessageDto): boolean {
    return !!(message.latitude && message.longitude);
  }

  /**
   * Verifica se a mensagem tem contato
   */
  static hasContact(message: MessageDto): boolean {
    return !!(message.contactName && message.contactName.trim());
  }

  /**
   * Mapeia mensagem legada para nova interface (compatibilidade)
   */
  static fromLegacyMessage(legacy: LegacyMessage): MessageDto {
    return {
      id: legacy.id || '',
      conversationId: legacy.conversationId || legacy.chatId || '',
      direction: legacy.isFromMe ? MessageDirection.Out : MessageDirection.In,
      type: this.mapLegacyType(legacy.type),
      createdAt: legacy.timestamp || legacy.createdAt || new Date().toISOString(),
      status: MessageStatus.Sent,
      fromMe: legacy.isFromMe || legacy.fromMe || false,
      isGroup: legacy.isGroup || false,
      text: legacy.body || legacy.content || legacy.text,
      mediaUrl: legacy.mediaUrl || legacy.file?.url,
      fileName: legacy.fileName || legacy.file?.name,
      mimeType: legacy.mimeType || legacy.file?.type,
      size: legacy.size || legacy.file?.size,
      fromNormalized: legacy.from || legacy.fromNormalized,
      chatId: legacy.chatId
    };
  }

  /**
   * Mapeia tipo legado para novo enum
   */
  private static mapLegacyType(type: string): MessageType {
    switch (type?.toLowerCase()) {
      case 'text':
      case 'chat':
        return MessageType.Text;
      case 'image':
        return MessageType.Image;
      case 'video':
        return MessageType.Video;
      case 'audio':
        return MessageType.Audio;
      case 'voice':
      case 'ptt':
        return MessageType.Voice;
      case 'document':
      case 'file':
        return MessageType.Document;
      case 'sticker':
        return MessageType.Sticker;
      case 'location':
        return MessageType.Location;
      case 'contact':
        return MessageType.Contact;
      case 'system':
        return MessageType.System;
      default:
        return MessageType.Text;
    }
  }
}

// === COMPATIBILIDADE LEGADA ===

/**
 * Interface para mensagens legadas durante conversão
 */
export interface LegacyMessage {
  id?: string;
  conversationId?: string;
  chatId?: string;
  isFromMe?: boolean;
  fromMe?: boolean;
  isGroup?: boolean;
  body?: string;
  content?: string;
  text?: string;
  timestamp?: string;
  createdAt?: string;
  type?: string;
  mediaUrl?: string;
  file?: {
    url?: string;
    name?: string;
    type?: string;
    size?: number;
  };
  fileName?: string;
  mimeType?: string;
  size?: number;
  from?: string;
  fromNormalized?: string;
}

/**
 * @deprecated Use MessageDto instead
 */
export interface WhatsAppMessage extends Partial<MessageDto> {
  content?: string;
}

/**
 * @deprecated Use MessageDto instead
 */
export interface ChatMessageDto extends Partial<MessageDto> {
  body?: string;
}
